libp2p/stream_muxer/mplex/mplex_stream.py

```diff
@@ -34,13 +37,33 @@ def __init__(self, stream_id: int, initiator: bool, mplex_conn: IMuxedConn) -> N
         self.local_closed = False
         self.remote_closed = False
         self.stream_lock = asyncio.Lock()
+        self._buf = None
 
-    async def read(self) -> bytes:
+    async def read(self, n: int = -1) -> bytes:
         """
         read messages associated with stream from buffer til end of file
+        :param n: number of bytes to read
         :return: bytes of input
         """
-        return await self.mplex_conn.read_buffer(self.stream_id)
+        if n == -1:
+            return await self.mplex_conn.read_buffer(self.stream_id)
+        return await self.read_bytes(n)
+
+    async def read_bytes(self, n: int) -> bytes:
+        if self._buf is None:
```

Comment:
This function could use some code comments because it isn't clear what we are doing with this other buffer and why.

Reply:
Sorry that I didn't clarify why I add this function. I intend to align `Mplx.read` with the one in [Go](https://github.com/libp2p/go-mplex/blob/62fe9554facaec3f80333b61ea8d694fe615705f/stream.go#L112-L140). This is because that the current API of `Stream` make us unable to read the exact number of bytes, which is necessary when we are reading `varint`s. `read_bytes` is a temporary workaround without breaking the existing `read()` behavior. I will replace `read_bytes` with `read` and modify the necessary places before changing this PR from the draft to open.

Reply:
Take a look at this alternate implementation and let me know what you think about the two with respect to readability.

```python
    def read_exactly(self, num_bytes: int) -> bytes:
        while len(self._buffer) < num_bytes:

            data = self._socket.recv(4096)

            if data == b"":
                raise OSError("Connection closed")

            self._buffer.extend(data)
        payload = self._buffer[:num_bytes]
        self._buffer = self._buffer[num_bytes:]
        return bytes(payload)
```

source: https://github.com/ethereum/trinity/blob/40c2b0425a612ae7d2fea184a5f1f85fd6d9b8bc/trinity/db/manager.py#L36-L47

Reply:
From the readability perspective, I think `read_exactly` is far better. I will change to this pattern. Thanks for the direction.